File: frontend/src/backtest/pooled.ts
import type { PooledMetric, PooledSlice, PooledStat } from "../api/hooks";

// Pure display logic for the POOLED panel (the rows.ts idiom: unit-tested, no React).
//
// Two rules run through all of it.
//
// **Nothing here invents a number.** It formats what the backend computed and it never derives a new
// statistic — in particular it does NOT subtract a null's median from the actual to show a "gap". That
// subtraction is the number a reader would quote, and a difference of two medians has no n and no
// dispersion of its own; rendering it as a headline would manufacture a confidence the run never
// measured. The four figures sit side by side and are read across the row, which is what the
// backend-authored banner tells the reader to do.
//
// **An insufficient n is marked, never hidden.** A slice below the gate still renders (pruning hides,
// it never vanishes) with its count visible and its tone quiet, because "we looked and there were
// three" is a finding and an absent row is not.

/** One `Stat` as text: the median, then its n. "—" when the statistic has no values at all. */
export function fmtStat(s: PooledStat | null | undefined): string {
  if (!s || s.n === 0 || s.median === null || s.median === undefined) return "—";
  const pct = (s.median * 100).toFixed(1);
  return `${s.median > 0 ? "+" : ""}${pct}%`;
}

/** The tone class for a formatted stat — positive/negative/neutral, or none when unknowable. */
export function statTone(s: PooledStat | null | undefined): string {
  if (!s || s.n === 0 || s.median === null || s.median === undefined) return "";
  return s.median > 0 ? "pos" : s.median < 0 ? "neg" : "";
}

/** The sample-size caption for a stat — "n=41". Empty when there is nothing to size. */
export function fmtN(s: PooledStat | null | undefined): string {
  if (!s || s.n === 0) return "";
  return `n=${s.n}`;
}

/** The four columns of a pooled metric, in the order they must be READ: the actual first, then the
 *  three things that make it interpretable. The labels carry the meaning, so they live here (one
 *  place) rather than being spelled out at each render site. */
export type MetricColumn = {
  id: "actual" | "excess" | "excess_median" | "vs_timing" | "vs_name";
  label: string;
  title: string;
  stat: PooledStat;
};

/** Does this report carry the basket's MEDIAN move (B)? A run from before it does not, and a column
 *  that is permanently "—" is noise — so an older report renders exactly as it always did. */
export function hasBasketMedian(m: PooledMetric): boolean {
  return (m.excess_vs_basket_median?.n ?? 0) > 0;
}

/** The excess columns, in the order the REPORT says to read them.
 *
 *  Two figures over one priced population. The MEAN is what an equal-weight basket position earned; the
 *  MEDIAN is what the typical member did. A thematic basket is right-skewed by construction, so one
 *  moonshot lifts the mean and leaves the median where it was — which means "beat the basket" can mean
 *  "missed the name that carried it". Which one leads is the backend's call (`headline_excess`), not
 *  this file's: a stored report has to say how it was meant to be read. */
export function excessColumns(m: PooledMetric, headline: string): MetricColumn[] {
  const mean: MetricColumn = {
    id: "excess",
    label: "vs equal-weight basket",
    title:
      "what an equal-weight basket position earned: the same window measured against the thesis basket's equal-weight MEAN move. A real portfolio answer — and one that a single runaway name in the basket can dominate.",
    stat: m.excess,
  };
  const med = m.excess_vs_basket_median;
  if (!med || med.n === 0) return [mean];
  const median: MetricColumn = {
    id: "excess_median",
    label: "vs the typical name",
    title:
      "the same window measured against the basket's MEDIAN move — what the typical member did, over exactly the same priced names as the mean beside it. The two part company when the basket is skewed, which is when the difference matters.",
    stat: med,
  };
  return headline === "mean" ? [mean, median] : [median, mean];
}

export function metricColumns(m: PooledMetric, headline = "mean"): MetricColumn[] {
  return [
    {
      id: "actual",
      label: "actual",
      title:
        "the realized median return over the hold window. On a universe assembled with hindsight this figure alone is not evidence — read it against the others beside it.",
      stat: m.actual,
    },
    ...excessColumns(m, headline),
    {
      id: "vs_timing",
      label: "vs timing null",
      title:
        "the SAME name entered on a randomly drawn session instead — the timing null. If the algorithm's timing carries no edge, the actual sits on top of this.",
      stat: m.vs_timing,
    },
    {
      id: "vs_name",
      label: "vs name null",
      title:
        "a randomly drawn OTHER name from the same roster on the same day — the name-selection null.",
      stat: m.vs_name,
    },
  ];
}

/** The slice table's key columns, in a fixed order so every row reads down the same lanes. The names
 *  are ALGORITHM dimensions by construction — there is no thesis column here and never will be (see
 *  the pooled report's own docstring: slicing outcomes by thesis tests the idea, not the timing). */
export const SLICE_COLUMNS: readonly { id: string; label: string; title: string }[] = [
  {
    id: "key1_source",
    label: "Key 1",
    title: "which trigger family turned the first lock — insider, activist, catalyst, revenue, theme",
  },
  {
    id: "confirmation_grade",
    label: "Confirmation",
    title: "the confirmation grade at the arm — volume-backed core vs momentum-only flip",
  },
  {
    id: "co_arm_bucket",
    label: "Co-arm",
    title:
      "how many members armed together that thesis-session: alone | 2-3 | 4-7 | 8+. Episodes in a burst are not independent observations.",
  },
  {
    id: "close_reason",
    label: "Closed by",
    title: "why the arm run ended",
  },
];

/** Slices ordered largest-n first, so the rows that can carry weight are read first; ties fall back to
 *  the key text for a stable order. A gated slice keeps its place in the list — it is sorted, not
 *  filtered. */
export function sortedSlices(slices: readonly PooledSlice[]): PooledSlice[] {
  return [...slices].sort(
    (a, b) => b.n - a.n || JSON.stringify(a.key ?? {}).localeCompare(JSON.stringify(b.key ?? {})),
  );
}

/** A mix map ({value: count}) as rows ordered by count, then name — the firing diagnostics' shape.
 *  Percentages are of the TOTAL episodes, which is why the total is passed in rather than summed: a
 *  mix whose values do not cover every episode must not silently renormalize to 100%. */
export function mixRows(
  mix: Record<string, number> | undefined,
  total: number,
): { label: string; n: number; pct: number | null }[] {
  const entries = Object.entries(mix ?? {});
  entries.sort((a, b) => b[1] - a[1] || a[0].localeCompare(b[0]));
  return entries.map(([label, n]) => ({
    label,
    n,
    pct: total > 0 ? n / total : null,
  }));
}

/** "82%" from 0.8231 — the co-arm headline. Null in, "—" out (never "NaN%"). */
export function fmtPct(x: number | null | undefined): string {
  if (x === null || x === undefined) return "—";
  return `${(x * 100).toFixed(0)}%`;
}

/** The quiet line that appears when the timing null had almost no sessions to draw from. A short
 *  window makes that null VACUOUS and the failure is silent otherwise: every draw prices a near-empty
 *  forward window, `vs_timing` comes back near the actual, and the row reads "the algorithm ties with
 *  chance" when it means "there was no chance to compare against". Null when there is nothing to warn
 *  about, so the line marks an exception rather than decorating every run (#7). */
export function timingNullCaveat(sessions: number, episodes: number): string | null {
  if (episodes === 0) return null;
  if (sessions >= 30) return null;
  return (
    `The timing null drew from ${sessions} session${sessions === 1 ? "" : "s"} — too few for the ` +
    `comparison to mean much. A vs-timing figure close to the actual here says the window was short, ` +
    `not that the timing is worthless.`
  );
}

File: backend/backtest/pooled.py
"""POOLED analysis — the algorithm as the unit, never the thesis.

The operator's reframe, and it is the safer frame as well as the truer one: the question is "globally, does
the timing algorithm behave sensibly", not "does it behave on this thesis". Pooling episodes across every
thesis tests the ALGORITHM. Slicing outcomes by thesis tests the THESIS, which is the leaderboard trap and
a #4 violation — the platform is opinionated about timing and deferential about the idea, and a screen that
ranked theses by realized return would quietly stop being deferential.

So Q4 is enforced STRUCTURALLY here rather than by convention: `PooledReport` has no thesis field, the
slices are by ALGORITHM only (Key-1 source x confirmation grade x co-arm bucket x close reason), and a test
walks the serialized payload looking for any thesis identifier. A per-thesis view is reachable only as a
single-thesis drill-down on the surface, from the run's own ledger — never from this report.

Two further honesty rules carried from the spec:

- **Every metric reports its two nulls beside it.** An absolute number on a hindsight universe is not
  evidence; the same number against random timing and random names on the same universe is.
- **Every metric reports its n and an `insufficient_n` flag**, and effective n is far below the episode
  count because 82% of episodes armed alongside a co-member. The co-arm slice is what makes that visible
  rather than assumed.

FIRING DIAGNOSTICS carry no outcome at all — counts, arm density, trigger mix, close reasons. They are the
half of this report that has no hindsight risk, and they are what catches a broken detector.
"""

from __future__ import annotations

from collections.abc import Sequence
from statistics import median
from typing import Any, Literal

from pydantic import BaseModel, Field

from backtest.nulls import EpisodeNulls
from replay.metrics import MIN_N
from replay.schema import Episode

SLICE_KEYS: tuple[str, ...] = ("key1_source", "confirmation_grade", "co_arm_bucket", "close_reason")

#: WHICH excess-over-basket is the headline — ONE place, and the report carries the answer so the surface
#: reads it rather than deciding for itself.
#:
#: The median, since B: the mean is what an equal-weight basket POSITION earned, and one moonshot in a
#: twenty-name basket moves it by a twentieth of its own move while leaving the typical member where it
#: was. "The algorithm beat the basket" off the mean can therefore mean "the algorithm missed the one name
#: that carried the theme". The median asks what the algorithm beat the TYPICAL name by, which is the same
#: question the name-selection null asks and is the one a timing platform has to answer. Both are always
#: reported; only the order and the emphasis change.
HeadlineExcess = Literal["median", "mean"]
HEADLINE_EXCESS: HeadlineExcess = "median"


class Stat(BaseModel):
    """One number with its sample size — never a bare figure."""

    n: int = 0
    median: float | None = None
    mean: float | None = None

    @classmethod
    def of(cls, values: Sequence[float | None]) -> "Stat":
        vals = [v for v in values if v is not None]
        if not vals:
            return cls()
        return cls(
            n=len(vals),
            median=round(median(vals), 6),
            mean=round(sum(vals) / len(vals), 6),
        )


class MetricWithNulls(BaseModel):
    """A pooled metric and the two things it must be read against.

    ``actual`` alone is uninterpretable on a universe assembled with hindsight. ``vs_timing`` and
    ``vs_name`` are the same statistic over the two null distributions, and ``excess`` strips the theme's
    own drift by measuring against the basket's equal-weight move over the identical window. A reader who
    quotes ``actual`` without them is quoting the operator's stock-picking, not the algorithm's timing.
    """

    name: str
    claim: str
    actual: Stat
    #: against the basket's equal-weight MEAN move — what an equal-weight basket position earned.
    #: UNCHANGED by B: same population, same arithmetic, same number.
    excess: Stat
    #: against the basket's MEDIAN move — what the TYPICAL member did. The headline since B; see
    #: `HEADLINE_EXCESS`. Empty on a report written before B, which is honest: those reports had no
    #: median to quote.
    excess_vs_basket_median: Stat = Field(default_factory=Stat)
    vs_timing: Stat
    vs_name: Stat
    insufficient_n: bool = True
    note: str = ""


class Slice(BaseModel):
    """One ALGORITHM slice. ``key`` names only algorithm dimensions — see SLICE_KEYS."""

    key: dict[str, str] = Field(default_factory=dict)
    n: int = 0
    actual: Stat = Field(default_factory=Stat)
    excess: Stat = Field(default_factory=Stat)
    excess_vs_basket_median: Stat = Field(default_factory=Stat)
    vs_timing: Stat = Field(default_factory=Stat)
    vs_name: Stat = Field(default_factory=Stat)
    insufficient_n: bool = True


class FiringDiagnostics(BaseModel):
    """Counts and mixes — NO outcome, so no hindsight risk. This is what catches a broken detector."""

    n_episodes: int = 0
    n_scoreable: int = 0
    key1_source_mix: dict[str, int] = Field(default_factory=dict)
    confirmation_grade_mix: dict[str, int] = Field(default_factory=dict)
    co_arm_bucket_mix: dict[str, int] = Field(default_factory=dict)
    close_reason_mix: dict[str, int] = Field(default_factory=dict)
    entry_grade_mix: dict[str, int] = Field(default_factory=dict)
    # the independence fact, stated as a number rather than left to be inferred
    pct_armed_with_a_co_member: float | None = None
    widest_single_session_group: int = 0
    # How many trading sessions the TIMING null had to draw from. Reported because a short window makes
    # that null vacuous and the failure is silent otherwise: on a four-session run there is roughly one
    # alternative entry date per episode, every draw prices a near-empty forward window, and `vs_timing`
    # comes back as a median of 0.0 on a handful of draws -- which reads like "the algorithm ties with
    # chance" when it actually means "there was no chance to compare against". The name null is unaffected
    # (it varies the NAME, not the date), which is why the two can disagree sharply on a short window.
    timing_candidate_sessions: int = 0


class PooledReport(BaseModel):
    """The whole pooled view. Deliberately carries NO thesis identifier of any kind — see the module
    docstring; a test walks this payload to keep it that way."""

    n_episodes: int = 0
    n_scoreable: int = 0
    null_draws: int = 0
    null_seed: str = ""
    min_n: int = MIN_N
    #: which excess-over-basket this report is to be READ on. On the artifact rather than in the surface's
    #: head, so a stored report says how it was meant to be read; a report written before B carries no
    #: field and the surface falls back to "mean", which is what those reports were read under.
    headline_excess: HeadlineExcess = HEADLINE_EXCESS
    banner: str = ""
    metrics: list[MetricWithNulls] = Field(default_factory=list)
    slices: list[Slice] = Field(default_factory=list)
    diagnostics: FiringDiagnostics = Field(default_factory=FiringDiagnostics)


def _mix(values: Sequence[Any]) -> dict[str, int]:
    out: dict[str, int] = {}
    for v in values:
        k = "none" if v is None else (v.value if hasattr(v, "value") else str(v))
        out[k] = out.get(k, 0) + 1
    return dict(sorted(out.items()))


def _slice_key(ep: Episode) -> dict[str, str]:
    def s(v: Any) -> str:
        return "none" if v is None else (v.value if hasattr(v, "value") else str(v))

    return {k: s(getattr(ep, k, None)) for k in SLICE_KEYS}


def build_report(
    episodes: list[Episode],
    nulls: list[EpisodeNulls],
    *,
    draws: int,
    seed: str,
    sessions: int = 0,
) -> PooledReport:
    """Pool the run's episodes and their nulls into the algorithm-level view."""
    by_key = {(n.thesis_id, n.security_id, n.arm_date): n for n in nulls}
    paired = [
        (ep, by_key[(ep.thesis_id, ep.security_id, ep.arm_date)])
        for ep in episodes
        if (ep.thesis_id, ep.security_id, ep.arm_date) in by_key
    ]
    scoreable = [(ep, n) for ep, n in paired if n.forward_return is not None]

    actual = [n.forward_return for _, n in scoreable]
    excess = [n.excess_return for _, n in scoreable]
    excess_median = [n.excess_return_vs_median for _, n in scoreable]
    timing = [d.forward_return for _, n in scoreable for d in n.timing]
    name = [d.forward_return for _, n in scoreable for d in n.name]

    metrics = [
        MetricWithNulls(
            name="arm_timing_forward_return",
            claim=(
                "timing (the flaw patched): the realized return over the hold window from when it armed, "
                "against the same name entered at random and against random other names on the same day"
            ),
            actual=Stat.of(actual),
            excess=Stat.of(excess),
            excess_vs_basket_median=Stat.of(excess_median),
            vs_timing=Stat.of(timing),
            vs_name=Stat.of(name),
            insufficient_n=len(actual) < MIN_N,
            note=(
                "read the four together: `actual` alone is a statement about a universe assembled with "
                "hindsight, and only the gaps are about the algorithm. The two excess figures are one "
                "population two ways -- against the basket's MEDIAN move (the typical member, the "
                "headline) and against its equal-weight MEAN (what an equal-weight basket position "
                "earned); they part company exactly when the basket is skewed, which is when the "
                "difference matters."
            ),
        )
    ]

    slices: dict[tuple, list] = {}
    for ep, n in scoreable:
        slices.setdefault(tuple(sorted(_slice_key(ep).items())), []).append(n)
    sliced = [
        Slice(
            key=dict(k),
            n=len(group),
            actual=Stat.of([g.forward_return for g in group]),
            excess=Stat.of([g.excess_return for g in group]),
            excess_vs_basket_median=Stat.of([g.excess_return_vs_median for g in group]),
            vs_timing=Stat.of([d.forward_return for g in group for d in g.timing]),
            vs_name=Stat.of([d.forward_return for g in group for d in g.name]),
            insufficient_n=len(group) < MIN_N,
        )
        for k, group in sorted(slices.items(), key=lambda kv: str(kv[0]))
    ]

    with_co = [ep for ep in episodes if ep.co_arm_count > 0]
    diagnostics = FiringDiagnostics(
        n_episodes=len(episodes),
        n_scoreable=len(scoreable),
        key1_source_mix=_mix([ep.key1_source for ep in episodes]),
        confirmation_grade_mix=_mix([ep.confirmation_grade for ep in episodes]),
        co_arm_bucket_mix=_mix([ep.co_arm_bucket for ep in episodes]),
        close_reason_mix=_mix([ep.close_reason for ep in episodes]),
        entry_grade_mix=_mix([ep.entry_grade for ep in episodes]),
        pct_armed_with_a_co_member=(round(len(with_co) / len(episodes), 4) if episodes else None),
        widest_single_session_group=max((ep.co_arm_count + 1 for ep in episodes), default=0),
        timing_candidate_sessions=sessions,
    )

    return PooledReport(
        n_episodes=len(episodes),
        n_scoreable=len(scoreable),
        null_draws=draws,
        null_seed=seed,
        metrics=metrics,
        slices=sliced,
        diagnostics=diagnostics,
        banner=(
            "POOLED ACROSS THESES — the unit is the ALGORITHM, never the thesis. Absolute returns on this "
            "universe are biased upward by selection (the baskets were authored in 2026 over names that "
            "had already moved), so read every metric against its two nulls and its excess-over-basket. "
            "Episodes are NOT independent: the co-arm slice shows how many arrived in bursts, and the "
            "effective n is far below the episode count."
        ),
    )
